tpm2/pcr_profile.go

```diff
@@ -87,85 +93,434 @@ type pcrProtectionProfileExtendPCRInstr struct {
+		p := &PCRProtectionProfileBranchPoint{
+			profile: b.profile,
+			parent:  &PCRProtectionProfileBranch{profile: b.profile},
+			instr:   new(pcrProtectionProfileBranchPointInstr)}
+		p.parent.currentBranchPoint = p
+		return p
+	}
+
+	if b.profile.failed() {
+		return b.parent
+	}
+
+	b.done = true
+	return b.parent
+}
+
+// EndBranch can be called when the caller is finished with this branch. Once this
+// has been called, no further changes can be made to this branch.
+//
+// It retuns a pointer to the branch point to which this branch was added.
+//
+// Note that this cannot be called on the root branch associated with a profile.
+func (b *PCRProtectionProfileBranch) EndBranch() *PCRProtectionProfileBranchPoint {
+	return b.endBranchInternal(1)
+}
+
+// AbortBranch can be called to remove this branch from the profile.
+//
+// It retuns a pointer to the branch point to which this branch was originally
+// added.
+//
+// Note that this cannot be called on the root branch associated with a profile.
+func (b *PCRProtectionProfileBranch) AbortBranch() *PCRProtectionProfileBranchPoint {
+	b.prepareToModifyBranch(1)
+
+	if b.parent == nil {
+		b.profile.fail(1, "cannot abort the root branch")
+		// Always return something to avoid having to check for nil
+		p := &PCRProtectionProfileBranchPoint{
+			profile: b.profile,
+			parent:  &PCRProtectionProfileBranch{profile: b.profile},
+			instr:   new(pcrProtectionProfileBranchPointInstr)}
+		p.parent.currentBranchPoint = p
+		return p
+	}
+
+	if b.profile.failed() {
+		return b.parent
+	}
+
+	b.done = true
+	b.parent.removeBranch(b)
+	return b.parent
+}
+
+// PCRProtectionProfile provides a way to define the PCR policy used to
+// protect a key sealed with SealKeyToTPM, and contains a root
+// PCRProtectionProfileBranch. It can generate compound PCR policies for
+// multiple conditions by making use of sub-branches.
 type PCRProtectionProfile struct {
-	instrs []pcrProtectionProfileInstr
+	root *PCRProtectionProfileBranch
+	err  error
 }
 
+// NewPCRProtectionProfile creates an empty PCR profile.
 func NewPCRProtectionProfile() *PCRProtectionProfile {
-	return &PCRProtectionProfile{}
+	profile := &PCRProtectionProfile{root: new(PCRProtectionProfileBranch)}
+	profile.root.profile = profile
+	return profile
 }
 
-// AddPCRValue adds the supplied value to this profile for the specified PCR. This action replaces any value set previously in this
-// profile. The function returns the same PCRProtectionProfile so that calls may be chained.
-func (p *PCRProtectionProfile) AddPCRValue(alg tpm2.HashAlgorithmId, pcr int, value tpm2.Digest) *PCRProtectionProfile {
-	if len(value) != alg.Size() {
-		panic("invalid digest length")
+func (p *PCRProtectionProfile) fail(skip int, msg string) {
+	if p.err != nil {
+		return
 	}
-	p.instrs = append(p.instrs, &pcrProtectionProfileAddPCRValueInstr{alg: alg, pcr: pcr, value: value})
+
+	_, file, line, _ := runtime.Caller(skip + 1)
```

Comment:
wouldn't filtering over pcr_profile.go in file achieve the same effect to skip without the hassle of it?

Reply:
I've kind of done that now, removing the skip argument and all of the *Internal functions, but the code skips over the qualified function names instead (everything that starts with "github.com/snapcore/secboot/tpm2.(*PCRProtectionProfile")